### PR TITLE
Closes #15

### LINKED ISSUE #15: Create dataset loader for JNLPBA
From http://www.geniaproject.org/shared-tasks/bionlp-jnlpba-shared-task-2004

### PR DESCRIPTION
**Pull request by benjaminbeilharz**
- **Name:** *JNLPBA*
- **Description:** *The BioNLP / JNLPBA Shared Task 2004 involves the identification and classification of technical terms referring to concepts of interest to biologists in the domain of molecular biology. The task was organized by GENIA Project based on the annotations of the GENIA Term corpus (version 3.02).*
- **Paper:** *[http://www.geniaproject.org/shared-tasks/bionlp-jnlpba-shared-task-2004](http://www.geniaproject.org/shared-tasks/bionlp-jnlpba-shared-task-2004)*
- **Data:** *[Training Data](http://www.nactem.ac.uk/GENIA/current/Shared-tasks/JNLPBA/Train/Genia4ERtraining.tar.gz)*, [Evaluation Data](http://www.nactem.ac.uk/GENIA/current/Shared-tasks/JNLPBA/Evaluation/Genia4ERtest.tar.gz)

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [ ] ~If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.~

### CODE REVIEW

**Review by ruisi-su — APPROVED** [review]
LGTM! Thanks @benjaminbeilharz for your contribution! 

note: this is an existing dataset in huggingface

**Comment by jason-fries** [comment]
@ruisi-su @benjaminbeilharz 
One question with this dataset in the biobio_kb schema: are you materializing a tag class *per token* or reconstituting spans of contiguous tokens to create the full entity mention?

**Comment by benjaminbeilharz** [comment]
@jason-fries it's done for each token as it is done in the original dataset, where they have lines of tokens and tags which are tab-separated. Sentences are split by an empty line (which does not effect the schema).

**Comment by jason-fries** [comment]
Thanks @benjaminbeilharz. For the `source` schema we would preserve the per-tag labeling, but for `bigbio_kb` entities are defined over spans so you'll need to convert internally before populating the schema.

**Comment by benjaminbeilharz** [comment]
@jason-fries I may need clarification, just to be sure. I have created the following:
```python
offset_start = 0
for token, tag in zip(sample["tokens"], sample["ner_tags"]):
    offset_start += len(token) + 1
    feature_dict["entities"].append(
                        {
                            "id": uid,
                            "offsets": [[offset_start, offset_start + len(token)]],
                            "text": [token],
                            "type": tag,
                            "normalized": [],
                        }
                    )
```
In the end, each sentence will consist of multiple entities, with a span and a tag type for each token right?
Or did I misunderstand the schema completely?

**Comment by ruisi-su** [comment]
@benjaminbeilharz I think what @jason-fries might mean (please correct me if I am wrong) for the `entities` field, skip the ones that are labeled '0' ("ner_tags: Array of tags, where 0 indicates no bio-entity mentioned, 1 signals the first token of a bio-entity and 2 the subsequent bio-entity tokens." from the HF [jnlpba](https://huggingface.co/datasets/jnlpba#annotations) dataset), and append only the non-zero tags with and note the offset _at the span level_. Then store the sentence in a separate `text` field. Something like [this](https://github.com/bigscience-workshop/biomedical/blob/master/examples/bc5cdr.py) but probably less complex.

For the sample:
`{'id': '1', 'tokens': ['IL-2', 'gene', 'expression', 'and', 'NF-kappa', 'B', 'activation', 'through', 'CD28', 'requires', 'reactive', 'oxygen', 'production', 'by', '5-lipoxygenase', '.'], 'ner_tags': [1, 2, 0, 0, 9, 10, 0, 0, 9, 0, 0, 0, 0, 0, 9, 0]}`
the entry (maybe) should look like this:
```
{'id': '1',
'entities': [ {'id':0, 'normalized':[], 'offsets': [[0, 4]], 'text': ['IL-2'], 'type': '1'}, {...}, {...} ],
'text': 'IL-2 gene expression and NF-kappa B activation through CD28 requires reactive oxygen production by 5-lipoxygenase .'
}

**Comment by galtay** [comment]
was just looking at this dataset as part of some BLURB statistics. 
synthesizing what jason and ruisi are saying, we need to merge the tokens into spans for the bigbio schema and map the integer types to their string values (https://huggingface.co/datasets/jnlpba/blob/main/jnlpba.py#L79)
```
                        datasets.features.ClassLabel(
                            names=[
                                "O",
                                "B-DNA",
                                "I-DNA",
                                "B-RNA",
                                "I-RNA",
                                "B-cell_line",
                                "I-cell_line",
                                "B-cell_type",
                                "I-cell_type",
                                "B-protein",
                                "I-protein",
                            ]
                        )
```



working with an abbreviated version of the example above, 
```
{'id': '1', 
'tokens': ['IL-2', 'gene', 'expression', 'and', 'NF-kappa', 'B', 'activation'], 
'ner_tags': [1, 2, 0, 0, 9, 10, 0]    # <- these index into the class names above
}
```

would translate to two entities in the bigbio schema (not 4 token level entities)
```
{'id': '1',
'entities': [
 {'id':0, 'normalized':[], 'offsets':...., 'text': ['IL-2 gene'], 'type': 'DNA'}, 
 {'id':1, 'normalized':[], 'offsets':..., 'text': ['NF-kappa B'], 'type': 'protein'}, {...} ],
'text': 'IL-2 gene expression and NF-kappa B activation through CD28 requires reactive oxygen production by 5-lipoxygenase .'
}
```

**Comment by benjaminbeilharz** [comment]
Thanks for the thorough explanation @galtay @ruisi-su 

I will implement this today and file another PR. Sorry for any inconveniences caused.

**Comment by ruisi-su** [comment]
@benjaminbeilharz thanks a lot for continuing to contribute and sorry for generating additional confusion. Please let me know when this PR is ready for review!
